### DISCUSSION TITLE
Age 7+ in the era of data: learning the tools for data analysis with EV3dev in 2024 & beyond

### DISCUSSION THREAD

**Discussion by Growflavor** [ev3dev/ev3dev#1633] [Ideas]
Work in Progress -- currently just an outline of links I would have liked to have known...

Please add your favorite EV3dev learning related links!
_____________________________________________________________________ 
**LEGO EV3 Sets**  (included links to US ebay **sold** listings to show current price points)

[EV3dev supports more hardware](https://www.ev3dev.org/docs/platform-comparison/) than just the armel EV3; however, I suggest starting with: 
1 x LEGO EV3 Core Set 45544 ([sold listings](https://www.ebay.com/sch/i.html?_from=R40&_nkw=EV3%20Core%20Set%2045544&_sacat=0&LH_Sold=1&LH_Complete=1&rt=nc&_udhi=300))

Then over time add other kits (annual gifts, 'the LEGO babysitter' ;-), etc.) to be able to easily do many of the diverse EV3 projects found online:
1 x LEGO EV3 Expansion Set 45560 ([sold listings](https://www.ebay.com/sch/i.html?_from=R40&_nkw=EV3%20Expansion%20Set%2045560&_sacat=0&LH_Complete=1&_udhi=200&LH_Sold=1&rt=nc))
1 x home set EV3 31313 ([sold listings](https://www.ebay.com/sch/i.html?_from=R40&_nkw=EV3%2031313&_sacat=0&LH_Complete=1&_udhi=250&LH_Sold=1&rt=nc))
multiple drive bases since it is much easier to switch the intelligent brick between them then it is to change variant

[LEGO City Cargo Train 60198 can also be programmed with Pybricks](https://pybricks.com/projects/sets/city/60198-cargo-train/speed-control/)

LEGO Spike Prime might become interesting for ML: https://www.rebeccalshen.com/projects/summer21

_____________________________________________________________________
**This post focuses on Python, however EV3dev is a Debian Linux OS so many programming languages may be used...**

https://www.ev3dev.org/docs/programming-languages/

![GEARSpythonMakecodeJS](https://github.com/user-attachments/assets/a15a57b1-be7b-4079-8010-573ce245ec27)

**EV3 Blocky & Python & Micropython**:
     [GEARSbot online](https://gears.aposteriori.com.sg/)
     Blocky for EV3dev2 python3 & pybricks-micropython -- note: only Blocks convert to Python
     great in combination with the [EV3dev2 API](https://ev3dev-lang.readthedocs.io/projects/python-ev3dev/en/stable/spec.html) & [pybricks-micropython API](https://pybricks.com/ev3-micropython/ev3devices.html)
     --simulator & bot configurator
     --some WRO challenges
     --source on GitHub: https://github.com/QuirkyCort/gears

**Online Pybricks Blocky & Micropython** for OTHER LEGO parts (trains, hubs, spikeprime,...): https://code.pybricks.com/

**EV3 basic**:
     https://sites.google.com/site/ev3basic/
     https://smallbasic-publicwebsite.azurewebsites.net/
     https://winthropdc.wordpress.com/2017/07/24/ev3basic-using-the-mindsensors-sensor-multiplexer-with-ev3-basic/

**EV3 JavaScript**:
     __Blocky & JavaScript__: Microsoft Makecode works with [OEM LEGO EV3 Firmware 1.10E](https://drive.google.com/file/d/1KsBAyvzOnBR5SBYPUT1u4aIQZwOvy77p/view?usp=sharing): https://makecode.mindstorms.com/
     
     EV3dev "Pathfind3r" printer project using EV3 as www server & .js & socket server: 
          [Pathfind3r](https://www.ev3dev.org/projects/2016/06/08/PATHFIND3R/)
          https://github.com/K2x141b/pathfind3r
          https://docs.google.com/presentation/d/17D7wMft5MAldGmK_YalZsMyZO5aGsjs3l2NbQHaelVc/edit#slide=id.p14

**EV3 C/C++**: https://github.com/c4ev3

https://www.ev3dev.org/docs/tutorials/

_____________________________________________________________________
**Age 7+**

One could start with the [core set & expansion set projects](https://education.lego.com/en-us/product-resources/mindstorms-ev3/teacher-resources/classroom-materials/) with [LEGO EV3classsroom Blocky](https://education.lego.com/en-us/downloads/mindstorms-ev3/software/).

--The ['drive base' a.k.a. 'robot educator'](https://pybricks.com/ev3-micropython/examples/robot_educator_basic.html) with the [gyro (PDF)](https://assets.education.lego.com/v3/assets/blt293eea581807678a/bltc961107e6587a579/5f8801dd3b19d21d4a6085bf/ev3-gyro-sensor-driving-base.pdf?locale=en-us) & [color sensor (PDF)](https://le-www-live-s.legocdn.com/sc/media/lessons/mindstorms-ev3/building-instructions/ev3-rem-color-sensor-down-driving-base-d30ed30610c3d6647d56e17bc64cf6e2.pdf) or [ultrasonic sensor (PDF)](https://le-www-live-s.legocdn.com/sc/media/lessons/mindstorms-ev3/building-instructions/ev3-ultrasonic-sensor-driving-base-61ffdfa461aee2470b8ddbeab16e2070.pdf) additions is one great versatile bot with which to do many explorations & experiments.

--various forms of 'Bumper Bots': [one example type (youtube)](https://youtu.be/woVlUWR4Fso?si=L90zBIUKPaeq2DKq), another is provided in the [Udemy course EV3 Python](https://www.udemy.com/course/ev3-python)

![BumperDriveBaseBot_opt](https://github.com/user-attachments/assets/f1823a13-9aa3-4a08-a6aa-7fc90c6977dc)

--[GyroBoy](https://pybricks.com/ev3-micropython/examples/gyro_boy.html) is more complicated, but a fantastic project.
https://github.com/Growflavor/EV3projects/tree/main/GyroBoyPython3
Port GyroBoy to Python3: https://github.com/Growflavor/EV3projects/commit/c7efdac77fa8006d3d1618fcb70488e359a04195
<img src="https://github.com/user-attachments/assets/8be60512-c021-4f79-a9e7-8968e2b9d760" width="400">

As interest in code 'behind the blocks' emerges, one can 
explore the provided [pybricks-micropython code for the LEGO coreset & expansion set projects](https://pybricks.com/ev3-micropython/examples/robot_educator_basic.html) &

Practice Python & MicroPython as either Blocky or Code accessible to young persons:
     --[GEARSbot online](https://gears.aposteriori.com.sg/)
     --source on GitHub: https://github.com/QuirkyCort/gears

[OpenRobertaLab also offers a Blocky interface](https://www.open-roberta.org/features/#EV3) for EV3dev as well as many other robotics platforms
    
[BlockPy online (think.cs.vt.edu/blockpy)
](https://think.cs.vt.edu/blockpy/blockpy/load?settings-display.instructor=true)    --not EV3dev specific, however python blocks & code convert both ways
    --programming 'turtle' can be fun for younger persons

[RobotLab has useful EV3dev API explanations with helpful code examples ](https://robolab.inf.tu-dresden.de/hardware/sensors/#gyro-sensor)

[Pybricks ecosystem](https://code.pybricks.com/) -- utilize multiple LEGO hub types:
     --online for LEGO® BOOST, City, Technic, and SPIKE® & via [VScode for EV3 MINDSTORMS®](https://pybricks.com/ev3-micropython/startrun.html).
     --Blocky programming now offered online

_____________________________________________________________________
**Age 10+**

**A nice interactive online (Colab) python3 practice/reminder/cheat sheet**: 
https://cs231n.github.io/python-numpy-tutorial/ in [Google CoLab Notebooks (online Jupyter notebook)](https://colab.research.google.com/github/cs231n/cs231n.github.io/blob/master/python-colab.ipynb)

**100 numpy exercises:** https://github.com/rougier/numpy-100

[VScode venv & pylance setup](https://pybricks.com/projects/tutorials/dev/tools/vscode/)
EV3dev VScode setup tutorial: https://youtu.be/TNXqizQTZhs

![VScodeEV3](https://github.com/user-attachments/assets/0785b9bd-ffc3-424c-847f-42cb9ea1340a)

[Reusing SSH Connections](https://www.ev3dev.org/docs/tutorials/reusing-ssh-connections/)

Control EV3 via RPyC in a VScode Jupyter notebook:
VScode has a jupyter note book extension so you can have a notebook open directly in VScode served from the [same .venv you use for the pylance interpreter](https://github.com/ev3dev/vscode-hello-python?tab=readme-ov-file#code-completion) & [use RPyC 5.3.1 to directly control your EV3 by notebook code cells](https://i.imgur.com/PWo6Qx2.png)

**Especially interesting EV3dev related projects (w.i.p.):**

--3Dprint LEGO compatible custom bricks: [LDraw2STL](https://github.com/kristov/ldraw2stl)

--[PID line follower MicroPython single color sensor](https://thecodingfun.com/2020/06/16/lego-mindstorms-ev3-pid-line-follower-code-by-using-micropython-2-0/)
--[PID line follower Python3 color & ultrasonic sensors](https://gist.github.com/CS2098/ecb3a078ed502c6a7d6e8d17dc095b48)

--socket server used in [Pathfind3r](https://www.ev3dev.org/projects/2016/06/08/PATHFIND3R/)
![Pathfind3r](https://github.com/user-attachments/assets/59d80ca8-0162-47b0-a3af-54a95d61e0d2)

--[µanthánō](https://leenahere.github.io/blog/workshop/) Blocky ML prototype

--[OpenRobertaLab's new neural network xNN](https://www.open-roberta.org/xnn/)
<img src="https://github.com/user-attachments/assets/68c7e9a6-e13d-44b5-a520-e4159e828e28" height ="250">
https://github.com/orgs/ev3dev/discussions/1646#discussioncomment-10526291

**Direct link to Artificial Neural Networks for EV3:**
https://lab.open-roberta.org/?loadSystem=ev3dev&extensions=nn
![Screenshot 2024-09-02 220114](https://github.com/user-attachments/assets/8c134c38-3d33-4016-b785-8b6aea74a7f2)

**Interesting EV3 projects (not ported to EV3dev):**

-- https://jkbrickworks.com/?s=ev3

--["2 Sensor PID Line Follower - The Most Efficient EV3 Line Follower"](https://youtu.be/ie2q1gr6Y7M?si=_3cQnvRpo4eRlQZ5)

--[Mindcub3r 3x3 cube solver](https://mindcuber.com/)  
solves traditional Rubik's cube in ~20 steps **onboard the EV3** = about 45 seconds
note: color sensor does not properly read colors of common 'speed cubes'
great for a person learning 3x3 cubes --at first they will 'lose' to the mindcub3r then eventually easily beat it
Mindcub3r also can make 'super flip' and other cube combinations for you to solve...swap in the speed cube in the same orientation as the Rubik's cube before mindcub3r starts the moves...

<img src="https://github.com/user-attachments/assets/ecb3cfec-4ba3-4641-af85-30dd743f922b" width=700>

[Make an online presentation about the EV3](https://docs.google.com/presentation/d/e/2PACX-1vRxtqOD65mFWdtD3ZxJWQavqv-gA8FQ1I7rEGKVTWRGzFQGYNucFPQ5-5TzofsjKLCFU25znzV8O41I/embed).

_____________________________________________________________________
**Age 12+**

--Port [EV3 Core Set projects from micropython](https://pybricks.com/ev3-micropython/examples/robot_educator_basic.html) to EV3devLangPython: [GyroBoy example](https://github.com/Growflavor/EV3projects/commit/c7efdac77fa8006d3d1618fcb70488e359a04195)

--Fork EV3 github project repositories
- clone repository to EV3: git clone --depth 1 <repo_URI>
- edit/push commits to one's fork via VScode 'remote repositories' extension
- 'git pull' changes to the folder of the EV3 clone

--Run tests & Build time critical loops with [EV3devFast](https://github.com/QuirkyCort/ev3dev-lang-python-fast)

--[Discuss project ideas with a LLM AI mentor such as ChatGPT](https://chatgpt.com/share/67ca3b82-ceb0-8008-8886-b22b17763c4b)

Spatial Math tutorials: https://github.com/jhavl/spatialmathematics

Practice with new features added to Python versions:
[3.7--Debian Buster image](https://docs.python.org/3.7/whatsnew/index.html)  
[3.8](https://docs.python.org/3.8/whatsnew/index.html)  
[3.9--Debian Bullseye image](https://docs.python.org/3.9/whatsnew/index.html)  
[3.10](https://docs.python.org/3.10/whatsnew/index.html)  
[3.11--Debian Bookworm & Alt Install on Debian Buster image](https://docs.python.org/3.11/whatsnew/index.html)

_____________________________________________________________________
**Age 15+**

engage the more complicated use of PiStorms, etc. & study/make python wrappers:
#1310 mindsensors NXTSumoEyes-V2 Driver
#1668 MS MUX with Pistorms
#1659 How to override/ignore EV3 Large Motor Max DPS
#1658 Connect RC servos with Mindsensors "8 Channel Servo Controller for NXT or EV3" 
[modified version of the YDLidar SDK and integrated it into an example project: YDLidarCar](https://www.ev3dev.org/projects/2019/01/05/YDLidarCar/)
#1391 BrickPi (Dexter Industries)

--customize [EV3 images in docker containers](https://hub.docker.com/u/ev3dev)
--use [Brickstrap to convert containers to SDcard images](https://github.com/ev3dev/brickstrap)

--[use docker to cross compile](https://www.ev3dev.org/docs/tutorials/using-docker-to-cross-compile/)

Some uses of python:
[QGIS](https://www.qgis.org/en/site/)
[PlanetarySystemStacker](https://github.com/Rolf-Hempel/PlanetarySystemStacker)
[Octoprint](https://community.octoprint.org/t/setting-up-octoprint-on-windows/383)
UDEMY Course: https://www.udemy.com/course/python-for-machine-learning-data-science-masterclass

**Comment by Growflavor** [comment]
1) an age 12+ next generation EV3 related project worth following -- **EVN (uses the Arduino platform)**: 
https://coresg.tech/evn/
https://evn.readthedocs.io/latest/getting-started/first-boot.html

2) **Consider ChatGPT as mentor aid to IT interested young persons:**

Since there is so much data online for the EV3 & python & JavaScript, large language models implemented via services such as ChatGPT have enough accuracy to enable a level of self driven creativity I had never anticipated.

To my surprise over the past year (age 12), using his basic knowledge of python, experimentation, & "mentoring" conversations with ChatGPT, my son has on his own assembled a number of successful, novel, intriguing, fully functional EV3 projects such as elaborate morse code signaling/speech between multiple EV3 using the IR sensors, python ANN projects, [created a website for his own spelling & vocabulary practice](https://sites.google.com/view/xwordpractice/white) that [uses a string comparison algorithm](https://chatgpt.com/share/67ca35d2-8620-8008-ad88-b1b3bbe67463), and others. 

For anyone interested in trying ChatGPT as a "mentor" with their children or students, I might be able to [share some of the recorded conversations to give a sense of how it can work](https://chatgpt.com/share/67ca3b68-d44c-8008-bb23-1122e26502f0).

One current project example, he emailed me today about a recent exploration of various EV3 track/maze runners:

"Yesterday I made the EV3 drive with our RC plane transmitter (joystick).
<img src="https://github.com/user-attachments/assets/55fed89c-0978-4303-8713-e97fc9c30803" width=700>

Today I made the track...I got the training data by driving myself with the joystick and then I trained the EV3.
...you can see the video of it driving by itself:
![V_20250305_205934_ES2optimize](https://github.com/user-attachments/assets/b36b397f-b074-4474-9cd4-72436809438d)
<img src="https://github.com/user-attachments/assets/d0435de9-eabf-4ea6-893e-adcd24aead11" width=700>

It can only turn left well.
Turning right is still a problem."

I speculate he was inspired by two projects we had looked at together awhile ago:
[Machine Learning on Arduino Uno was a Good Idea](https://www.youtube.com/watch?v=PdSDhdciSpE)
[LEGO EV3 Maze-Driving Robot](https://www.instructables.com/AI-in-LEGO-EV3-Maze-Driving-Robot/)

Another, while not yet connected to an EV3 but at least useful to compare the method with the EV3 ML LineFollower [µanthánō](https://leenahere.github.io/blog/workshop/) , he found & implemented a project where a NN is trained to interpret chalkboard input:

<img src="https://github.com/user-attachments/assets/035a6298-e46e-4673-acf9-8d7608561b0b" width=700>

**Comment by Growflavor** [comment]
RE: #1391

https://www.ev3dev.org/docs/platform-comparison/

I've been thinking about whether a BrickPi might be a useful complement alongside our OEM LEGO EV3 bricks since it looks like parts **might** still be available ([appears DI offered it it with a Pi 3B+ in 2020](https://www.dexterindustries.com/brickpi-2020/)): 
https://www.dexterindustries.com/brickpi-core/
https://www.adafruit.com/product/3775

@dlech would value your thoughts on the practical reality of working with a BrickPi relative to the OEM LEGO EV3 brick, and whether it substantially extends the capability/speed of what one can do with EV3dev -- e.g. down the road we've been thinking that some of the LIDAR mapping or ML projects might be fun to try to replicate?

As one limited example of a practical difference, I see from issue #1391 the 'reality' of the difference with identifying sensors/motors.

I am a bit surprised dexterindustries gave it no buttons...
& am I understanding correctly that the BrickPi has its own firmware & then EV3dev runs on the Pi 3B+ & communicates with the BrickPi board?

Thanks!

**Reply by dlech** [reply]
Personally, I don't have the time and energy anymore to try to support more than EV3 with ev3dev. At one point in time BrickPi3 was quite popular, but it has been a long time since I've seen anyone asking questions about it, so I don't think there is much of an active user base anymore. 

>  am I understanding correctly that the BrickPi has its own firmware & then EV3dev runs on the Pi 3B+ & communicates with the BrickPi board?

Yes, this is correct. And since everything has to communicate over a UART, it makes motor control not work as well as on EV3. If I could start over, I would write a new firmware for BrickPi3 that ran Pybricks on the BrickPi3 itself for better control. And this would also make it usable without needing an RPi.

**Reply by Growflavor** [reply]
Thank you for your insights (& remarkable contribution to EV3dev)...Regarding EV3dev then, we'll stick with just our (5 ;-) EV3 then.

especially with the ["Debian Buster beta"](https://github.com/orgs/ev3dev/discussions/1634) the EV3 continues to be a truly fantastic & very broad segment of our current IT learning platforms (EV3dev/[VScode](https://www.youtube.com/watch?v=TNXqizQTZhs);[venv](https://github.com/ev3dev/vscode-hello-python?tab=readme-ov-file#code-completion);[jupyter](https://photos.app.goo.gl/cG35DffEMyfgeebJ9)/[CoLab](https://colab.research.google.com/github/cs231n/cs231n.github.io/blob/master/python-colab.ipynb), [3D printing](https://www.thingiverse.com/growflavor/designs) / [LDraw2STL](https://github.com/kristov/ldraw2stl), [CAM design](https://photos.app.goo.gl/xLgc6QyyzwG3WtBw7), [Arduino kits](https://us.elegoo.com/collections/arduino-kits/products/elegoo-uno-most-complete-starter-kit), [RC planes](https://www.tapatalk.com/groups/picasim/r-w-i-p-research-work-in-progress-picasim-modding--t404.html), [www design](https://sites.google.com/view/xwordpractice/white))...

I never imagined a young person could learn so many 'effective' lifelong IT/manufacturing design workflows so quickly, in such a 'fun', rewarding, sustained, & experimental way, nor the remarkable usefulness of [a young person's conversations with LLM 'AI mentors' ](https://chatgpt.com/share/67ca3cf2-1978-8008-aaa5-5d2bb5f5aa97) & huge amount of excellent [youtube](https://youtube.com/playlist?list=PL8bk60c_Nfo-VwYr7kIDB_23Sm-TE906l&si=NPNOzvpwksvLYiTp)/instructibles/[github tutorials/projects](https://www.ev3dev.org/projects/) available for assistance/project templates.

I'll keep an eye instead on the Spike Prime hubs, given the [Pybricks ecosystem](https://code.pybricks.com/) also [works with our LEGO City train hub](https://pybricks.com/learn/getting-started/what-do-you-need/).  Maybe at some point we'll add a Spike Prime hub & [some "qikeasy" adapters to use EV3 parts with it](https://www.qikeasy.com/prod/qikeasyadapter/)...
